npm/pkg/dataplane/dataplane.go

```diff
@@ -366,52 +394,151 @@ func (dp *DataPlane) applyDataPlaneNow(context string) error {
+			// remove from queue on success
+			klog.Infof("[DataPlane] [%s] added policy successfully one at a time. policyKey: %s", context, netPol.PolicyKey)
+			dp.netPolQueue.delete(netPol.PolicyKey)
+		} else {
+			// keep in queue on failure
+			klog.Errorf("[DataPlane] [%s] failed to add policy one at a time. policyKey: %s. err: %s", context, netPol.PolicyKey, err.Error())
+			metrics.SendErrorLogAndMetric(util.DaemonDataplaneID, "[DataPlane] [%s] failed to add policy one at a time. %s. err: %s", context, netPol.PolicyKey, err.Error())
+		}
+	}
+}
+
+func (dp *DataPlane) addPolicies(netPols []*policies.NPMNetworkPolicy) error {
+	if !dp.netPolInBackground && len(netPols) != 1 {
+		klog.Errorf("[DataPlane] expected to have one NetPol in dp.addPolicies() since dp.netPolInBackground == false")
+		metrics.SendErrorLogAndMetric(util.DaemonDataplaneID, "[DataPlane] expected to have one NetPol in dp.addPolicies() since dp.netPolInBackground == false")
+		return ErrIncorrectNumberOfNetPols
+	}
+
+	if len(netPols) == 0 {
+		klog.Infof("[DataPlane] expected to have at least one NetPol in dp.addPolicies()")
+		return nil
+	}
+
+	// 1. Add IPSets and apply for each NetPol.
+	// Apply IPSets after each NetworkPolicy unless ApplyInBackground=true and we're in the bootup phase (only happens for Windows currently)
+	wasInBootupPhase := true
+	for _, netPol := range netPols {
+		// Create and add references for Selector IPSets first
+		err := dp.createIPSetsAndReferences(netPol.AllPodSelectorIPSets(), netPol.PolicyKey, ipsets.SelectorType)
 		if err != nil {
-			return err
+			klog.Infof("[DataPlane] error while adding Selector IPSet references: %s", err.Error())
+			return fmt.Errorf("[DataPlane] error while adding Selector IPSet references: %w", err)
 		}
-	} else {
-		err = dp.applyDataPlaneNow(contextAddNetPol)
+
+		// Create and add references for Rule IPSets
+		err = dp.createIPSetsAndReferences(netPol.RuleIPSets, netPol.PolicyKey, ipsets.NetPolType)
 		if err != nil {
-			return err
+			klog.Infof("[DataPlane] error while adding Rule IPSet references: %s", err.Error())
+			return fmt.Errorf("[DataPlane] error while adding Rule IPSet references: %w", err)
 		}
 
-		endpointList, err = dp.getEndpointsToApplyPolicy(policy)
+		if dp.inBootupPhase() {
+			// This branch can only be taken in Windows.
+			// During bootup phase, the Pod controller will not be running.
+			// We don't need to worry about adding Policies to Endpoints, so we don't need IPSets in the kernel yet.
+			// Ideally, we get all NetworkPolicies in the cache before the Pod controller starts
+			err = dp.incrementBatchAndApplyIfNeeded(contextAddNetPol)
+			if err != nil {
+				return err
+			}
+		} else {
+			// This branch is always taken in Linux.
+			wasInBootupPhase = false
+			err = dp.applyDataPlaneNow(contextAddNetPol)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// 2. get Endpoints
+	var err error
+	var endpointList map[string]string
+	if !dp.inBootupPhase() {
```

Comment:
Is this condition correct ? the only condition for wasInBootPhase to be true is if dp.inBootupPhase() returns true. 

here you are checking the opposite and wasInBootPhase , i think we will never hit this condition ? unless i am missing something ?

Reply:
The idea is to protect against a race:
1. DP in bootup phase.
2. dp.incrementBatchAndApplyIfNeeded() is called in last iteration of above loop.
3. dp.FinishBootupPhase() in another thread

Few more details in comment below